### ISSUE TITLE
no-extra-parens:  add option that would allow extra parans in math expressions while concatenating strings.

### ISSUE THREAD

**Issue by sebenik** [eslint/eslint#5584]
**------------------------**
**False bug report!**
**------------------------**

**What version of ESLint are you using?**
v2.4.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
```
{
	"env" : {
		"browser" : true, // browser global variables.
		"node" : true, // Node.js global variables and Node.js-specific rules.
	},

	"globals" : {
		"$" : true,
		"google" : true
	},

	"rules" : {
		"comma-dangle" : 2, // disallow trailing commas in object literals
		"no-cond-assign" : 1, // disallow assignment in conditional expressions
		"no-dupe-args" : 1, // disallow duplicate arguments in function
		"no-dupe-keys" : 2, // disallow duplicate keys when creating object literals
		"no-duplicate-case" : 2, // disallow a duplicate case label
		"no-empty" : 1, // disallow empty statements
		"no-ex-assign" : 2, // disallow assigning to the exception in a catch block
		"no-extra-boolean-cast" : 2, // disallow double-negation boolean casts in a boolean context
		"no-extra-parens" : 1, // disallow unnecessary parentheses (off by default)
		"no-extra-semi" : 2, // disallow unnecessary semicolons
		"no-func-assign" : 2, // disallow overwriting functions written as function declarations
		"no-inner-declarations" : 1, // disallow function or variable declarations in nested blocks
		"no-invalid-regexp" : 2, // disallow invalid regular expression strings in the RegExp constructor
		"no-irregular-whitespace" : 2, // disallow irregular whitespace outside of strings and comments
		"no-obj-calls" : 2, // disallow the use of object properties of the global object (Math and JSON) as functions
		"no-sparse-arrays" : 2, // disallow sparse arrays
		"no-unexpected-multiline" : 2, // Avoid code that looks like two expressions but is actually one
		"no-unreachable" : 1, // disallow unreachable statements after a return, throw, continue, or break statement
		"valid-typeof" : 1, // Ensure that the results of typeof are compared against a valid string

		"accessor-pairs" : 1, // Enforces getter/setter pairs in objects
		"curly" : [2, "all"], // specify curly brace conventions for all control statements
		"default-case" : 1, // require default case in switch statements (off by default)
		"dot-notation" : 2, // encourages use of dot notation whenever possible
		"eqeqeq" : 2, // require the use of === and !==
		"no-alert" : 2, // disallow the use of alert, confirm, and prompt
		"no-case-declarations" : 2, // disallow lexical declarations in case clauses
		"no-else-return" : 2, // disallow else after a return in an if (off by default)
		"no-empty-function" : 2, // disallow use of empty functions
		"no-empty-pattern" : 2, // disallow use of empty destructuring patterns
		"no-eq-null" : 2, // disallow comparisons to null without a type-checking operator (off by default)
		"no-eval" : 2, // disallow use of eval()
		"no-fallthrough" : 2, // disallow fallthrough of case statements
		"no-floating-decimal" : 2, // disallow the use of leading or trailing decimal points in numeric literals (off by default)
		"no-implied-eval" : 2, // disallow use of eval()-like methods
		"no-labels" : 2, // disallow use of labeled statements
		"no-lone-blocks" : 2, // disallow unnecessary nested blocks
		"no-loop-func" : 2, // disallow creation of functions within loops
		"no-multi-spaces" : 2, // disallow use of multiple spaces
		"no-multi-str" : 2, // disallow use of multiline strings
		"no-native-reassign" : 2, // disallow reassignments of native objects
		"no-new" : 2, // disallow use of new operator when not part of the assignment or comparison
		"no-new-func" : 2, // disallow use of new operator for Function object
		"no-new-wrappers" : 2, // disallows creating new instances of String, Number, and Boolean
		"no-octal" : 2, // disallow use of octal literals
		"no-octal-escape" : 2, // disallow use of octal escape sequences in string literals, such as var foo = "Copyright \251";
		"no-redeclare" : 2, // disallow declaring the same variable more then once
		"no-return-assign" : 1, // disallow use of assignment in return statement
		"no-script-url" : 2, // disallow use of javascript: urls.
		"no-self-assign" : 2, // disallow assignments where both sides are exactly the same
		"no-sequences" : 2, // disallow use of comma operator
		"no-unmodified-loop-condition" : 1, // disallow unmodified conditions of loops
		"no-unused-expressions" : 1, // disallow usage of expressions in statement position
		"no-unused-labels" : 2, // disallow unused labels
		"no-useless-call" : 2, // disallow unnecessary .call() and .apply()
		"no-useless-concat" : 2, // disallow unnecessary concatenation of literals or template literals
		"no-void" : 2, // disallow use of void operator (off by default)
		"no-warning-comments" : 1, // disallow usage of configurable warning terms in comments, e.g. TODO or FIXME (off by default)
		"no-with" : 2, // disallow use of the with statement
		"radix" : 2, // require use of the second argument for parseInt() (off by default)
		"yoda" : 2, // require or disallow Yoda conditions

		"strict" : [2, "safe"], // controls location of Use Strict Directives

		"no-catch-shadow" : 2, // disallow the catch clause parameter name being the same as a variable in the outer scope (off by default in the node environment)
		"no-delete-var" : 2, // disallow deletion of variables
		"no-label-var" : 2, // disallow labels that share a name with a variable
		"no-undef" : 2, // disallow use of undeclared variables unless mentioned in a /*global */ block
		"no-undef-init" : 2, // disallow use of undefined when initializing variables
		"no-undefined" : 2, // disallow use of undefined variable (off by default)
		"no-unused-vars" : 2, // disallow declaration of variables that are not used in the code
		"no-use-before-define" : 2, // disallow use of variables before they are defined

		"array-bracket-spacing" : [2, "never"], // enforce spacing inside array brackets
		"block-spacing" : 2, // disallow or enforce spaces inside of single line blocks
		"brace-style" : [2, "1tbs", {
				"allowSingleLine" : true
			}
		], // enforce one true brace style (off by default)
		"comma-spacing" : [2, {
				"before" : false,
				"after" : true
			}
		], // enforce spacing before and after comma
		"computed-property-spacing" : [2, "never"], // require or disallow padding inside computed properties
		"consistent-this" : [2, "self"], // enforces consistent naming when capturing the current execution context (off by default)
		"eol-last" : 2, // enforce newline at the end of file, with no multiple empty lines
		"indent" : [2, 2, {
				"SwitchCase" : 1
			}
		], // specify tab or space width for your code
		"key-spacing" : 2, // enforces spacing between keys and values in object literal properties
		"keyword-spacing" : 2, // enforce spacing before and after keywords
		"linebreak-style" : 2, // disallow mixed ‘LF’ and ‘CRLF’ as linebreaks
		"new-parens" : 2, // disallow the omission of parentheses when invoking a constructor with no arguments
		"newline-per-chained-call" : [2, {
				"ignoreChainWithDepth" : 3
			}
		], // enforce newline after each call when chaining the calls
		"no-array-constructor" : 2, // disallow use of the Array constructor
		"no-lonely-if" : 2, // disallow if as the only statement in an else block
		"no-mixed-spaces-and-tabs" : 2, // disallow mixed spaces and tabs for indentation
		"no-nested-ternary" : 2, // disallow nested ternary expressions
		"no-new-object" : 2, // disallow the use of the Object constructor
		"no-plusplus" : [2, {
				allowForLoopAfterthoughts : true
			}
		], // disallow use of unary operators, ++ and --
		"no-spaced-func" : 2, // disallow space between function identifier and application
		"no-trailing-spaces" : 2, // disallow trailing whitespace at the end of lines
		"no-unneeded-ternary" : 2, // disallow the use of ternary operators when a simpler alternative exists
		"no-whitespace-before-property" : 2, // disallow whitespace before properties
		"object-curly-spacing" : 2, // require or disallow padding inside curly braces
		"operator-linebreak" : [2, "before"], // enforce operators to be placed before or after line breaks
		"quote-props" : [2, "always"], // require quotes around object literal property names
		"quotes" : [2, "double", "avoid-escape"], // specify whether backticks, double or single quotes should be used
		"semi" : 2, // require or disallow use of semicolons instead of ASI
		"semi-spacing" : [2, {
				"before" : false,
				"after" : true
			}
		], // enforce spacing before and after semicolons
		"space-before-blocks" : 2, // require or disallow a space before blocks
		"space-before-function-paren" : [2, "never"], // require or disallow a space before function opening parenthesis
		"space-infix-ops" : [2, {
				"int32Hint" : false
			}
		], // require spaces around operators
		"spaced-comment" : 2 // require or disallow a space immediately following the // or /* in a comment
	}
}
```

**What did you do? Please include the actual source code causing the issue.**
```var foo = "bar " + (1 + 2) + " baz"```

**What did you expect to happen?**
I expect that foo equals ```"bar 3 baz"```

**What actually happened? Please include the actual, raw output from ESLint.**
Output from ESLint for this is: "Gratuitous parentheses around expression. (no-extra-parens) "
But this is wrong because without parenthesis foo equals ```"bar 12 baz"```

**Comment by platinumazure** [comment]
Can't reproduce in the demo (but I'm not sure the demo is at latest ESLint). If this can be reproduced, though, I'd say it's a bug and that the rule should be augmented so that it doesn't care about parens for string concatenation.

**Comment by sebenik** [comment]
I tested again and i was wrong before.

Within string concatenation, summation and subtraction requires parenthesis and ESLint **doesn't report an error**  ```"Gratuitous parentheses around expression. (no-extra-parens) "```
Error is **only shown** on multiplication and division.

**Comment by platinumazure** [comment]
@sebenik In that case, that's as expected. Multiplication and division have higher precedence than string concatenation, so there's no need to have parentheses.

I could maybe get behind an enhancement to the rule (behind a new rule option) which would allow extra parens around mathematical expressions in string concatenation. Then again, though, you can also work around this by either not using parens or by storing the mathematical result in a variable before doing the string concatenation.

**Comment by sebenik** [comment]
Yes this works as expected although additional option for this rule that would allow extra parans in math expressions while string concatenation would be appreciated.

**Comment by BYK** [comment]
Should we file a new issue or update this one to reflect the actual/new enhancement request?

**Comment by platinumazure** [comment]
@BYK The original request is invalid, so I could see it either way.

@sebenik If you agree that the original enhancement request is no longer valid, could you please either (a) edit the title and original post of the issue to reflect your new enhancement request, or (b) close this issue and open a new one with your new enhancement request? Thanks!

**Comment by kaicataldo** [comment]
Closing this because we haven't made any progress in months. Thanks for contributing!
